refactor: clean up error messages

dshemetov opened the pull request:
### Checklist

Please:

-   [x] Make sure this PR is against "dev", not "main".
-   [x] Request a review from one of the current epiprocess main reviewers:
        brookslogan, nmdefries.
-   [x] Makes sure to bump the version number in `DESCRIPTION` and `NEWS.md`.
        Always increment the patch version number (the third number), unless you are
        making a release PR from dev to main, in which case increment the minor
        version number (the second number).
-   [x] Describe changes made in NEWS.md, making sure breaking changes
        (backwards-incompatible changes to the documented interface) are noted.
        Collect the changes under the next release number (e.g. if you are on
        0.7.2, then write your changes under the 0.8 heading).

### Change explanations for reviewer

- Removed `Abort` and `Warn`, replaced with direct calls to `cli` functions
- Replaced many custom arg validation statements with `checkmate` equivalents
- Updated test regexp expectations

### Magic GitHub syntax to mark associated Issue(s) as resolved when this is merged into the default branch

-   Resolves #380

dajmcdon commented on R/archive.R:
```diff
@@ -278,26 +246,9 @@ epi_archive <-
-        }
+        assert_data_frame(x)
+        assert_subset(c("geo_value", "time_value", "version"), names(x))
+        assert(!anyNaN(x$version))
```
I think this last one is wrong (should be `anyNA()`).


Generally, I'm not sure I like the messages produced by this package. There's of course a balance between offloading this stuff to well-tested functions and getting everything I want, but these seem too low-tech. For example:

``` r
library(checkmate)
x <- 1:10
names(x) <- letters[1:10]
assert_subset(c("geo_value", "time_value", "version"), names(x))
#> Error: Assertion on 'c("geo_value", "time_value", "version")' failed: Must be a subset of {'a','b','c','d','e','f','g','h','i','j'}, but has additional elements {'geo_value','time_value','version'}.

assert_data_frame(x)
#> Error: Assertion on 'x' failed: Must be of type 'data.frame', not 'integer'.
version <- c(1:5, NA)
assert(!anyNaN(version))
assert(!anyNA(version))
#> Error: Assertion on 'anyNA(version)' failed: FALSE.
```

The messages aren't wrapped or formatted as they were before. And the messages aren't nearly as clear. It's worth examining https://style.tidyverse.org/error-messages.html

dshemetov replied:
> should be `anyNA()`

Ah yea, in the idioms of this packages, it's `anyMissing()`. I can change that.

> these seem too low-tech
> the messages aren't nearly as clear

Seems like a matter of preference, so it's hard for me to know what you don't like about these. Custom error messages can always be done like this
```r
if (test_subset(c("geo_value", "time_value", "version"), names(x))) cli_abort("custom message")
```
Like you said, the off-loading to well tested functions is exactly what I like here. It's wild to me how much of our unit testing is the same "did this input type error correctly throw?" checks over and over. To me that's just downstream of how squirmy type checking can get in R. This package gives me type check tests I can trust, which frees up energy for testing the higher level logic of the package.

dajmcdon replied:
Maybe better then to refactor the `arg_is_*()` type functions as you did in epipredict to use `{checkmate}`. That way you mostly get better messaging. 

In this particular case, to better mimic what the original messages were saying (though with better formatting), I'd like to see something like:
``` r
hardhat::validate_column_names(data.frame(geo_value = 1:5, time_value = 1:5), c("geo_value", "time_value", "version"))
#> Error in `hardhat::validate_column_names()`:
#> ! The following required columns are missing: 'version'.
```

<sup>Created on 2024-02-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>
```

dshemetov commented on R/archive.R:
```diff
@@ -278,26 +251,11 @@ epi_archive <-
-            class = "epiprocess__version_values_must_not_be_na"
-          )
-        }
+        assert_data_frame(x)
```
The message from this function is similar:
```R
r$> x = 5

r$> checkmate::assert_data_frame(x)
Error: Assertion on 'x' failed: Must be of type 'data.frame', not 'double'.
```

dshemetov commented on R/archive.R:
```diff
@@ -312,24 +270,21 @@ epi_archive <-
-          !rlang::is_null(compactify)) {
-          Abort("compactify must be boolean or null.")
         }
+        assert_logical(compactify, len = 1, null.ok = TRUE)
```
The message from this function is similar:
```R
r$> compactify = 5

r$> checkmate::assert_logical(compactify, len 
    = 1, null.ok = TRUE)
Error: Assertion on 'compactify' failed: Must be of type 'logical' (or 'NULL'), not 'double'.
```

dshemetov commented on R/archive.R:
```diff
@@ -493,24 +450,20 @@ epi_archive <-
-        }
+        assert_set_equal(class(max_version), class(self$DT$version))
+        assert_set_equal(typeof(max_version), typeof(self$DT$version))
+        assert_scalar(max_version, na.ok = FALSE)
```
The messaging here is similar:
```R
r$> max_version = c(5, 5)

r$> checkmate::assert_scalar(max_version, na.o
    k = FALSE)
Error: Assertion on 'max_version' failed: Must have length 1.
```

dshemetov commented on R/archive.R:
```diff
@@ -621,25 +573,17 @@ epi_archive <-
-        }
+        assert_set_equal(class(max_version), class(self$DT$version))
+        assert_set_equal(typeof(max_version), typeof(self$DT$version))
+        assert_scalar(max_version, na.ok = FALSE)
```
The messaging here is similar:
```R
r$> max_version = c(5, 5)

r$> checkmate::assert_scalar(max_version, na.o
    k = FALSE)
Error: Assertion on 'max_version' failed: Must have length 1.
```

dshemetov commented on R/correlation.R:
```diff
@@ -78,12 +78,11 @@
                     method = c("pearson", "kendall", "spearman")) {
-  # Check we have an `epi_df` object
-  if (!inherits(x, "epi_df")) Abort("`x` must be of class `epi_df`.")
+  assert_class(x, "epi_df")
```
```R
r$> assert_class(x, "epi_df")
Error: Assertion on 'x' failed: Must inherit from class 'epi_df', but has class 'numeric'.
```

dshemetov commented on R/epi_df.R:
```diff
@@ -114,14 +114,9 @@ NULL
-  if (!is.data.frame(x)) {
-    Abort("`x` must be a data frame.")
-  }
+  assert_data_frame(x)
```
```R
r$>   assert_data_frame(x)
Error: Assertion on 'x' failed: Must be of type 'data.frame', not 'double'.
```

dshemetov commented on R/epi_df.R:
```diff
@@ -114,14 +114,9 @@ NULL
-    Abort("`x` must be a data frame.")
-  }
+  assert_data_frame(x)
+  assert_list(additional_metadata)
```
r$> assert_list(additional_metadata)
Error: Assertion on 'additional_metadata' failed: Must be of type 'list', not 'double'.

dshemetov commented on R/grouped_epi_archive.R:
```diff
@@ -53,41 +53,22 @@ grouped_epi_archive <-
-            epiprocess__DT_names = names(ungrouped$DT)
-          )
-        }
+        assert_class(ungrouped, "epi_archive")
```
```R
r$> ungrouped = 5

r$> assert_class(ungrouped, "epi_archive")
Error: Assertion on 'ungrouped' failed: Must inherit from class 'epi_archive', but has class 'numeric'.
```

dshemetov commented on R/grouped_epi_archive.R:
```diff
@@ -53,41 +53,22 @@ grouped_epi_archive <-
-          )
-        }
+        assert_class(ungrouped, "epi_archive")
+        assert_character(vars)
```
```R
r$> vars = 5

r$> assert_character(vars)
Error: Assertion on 'vars' failed: Must be of type 'character', not 'double'.
```

dshemetov commented on R/grouped_epi_archive.R:
```diff
@@ -53,41 +53,22 @@ grouped_epi_archive <-
-          )
+          cli_abort("`version` has a special interpretation and cannot be used by itself as a grouping variable")
         }
+        assert_logical(drop)
```
```R
r$> assert_logical(5, len = 1)
Error: Assertion on '5' failed: Must be of type 'logical', not 'double'.
```

dshemetov replied:
Do you feel like the previous one was clearer? Is it the `rlang::last_trace()` message that you wish was there?

dshemetov commented on R/grouped_epi_archive.R:
```diff
@@ -230,32 +209,23 @@ grouped_epi_archive <-
-        } else if (any(ref_time_values > private$ungrouped$versions_end)) {
-          Abort("All `ref_time_values` must be `<=` the `versions_end`.")
         } else {
+          assert_numeric(ref_time_values, upper = private$ungrouped$versions_end, min.len = 1L, null.ok = FALSE, any.missing = FALSE)
```
```R
r$> ref_time_values = Date()

r$> assert_numeric(ref_time_values, 
    upper = as.Date("2020-03-05"), min.len = 1
    L, null.ok = FALSE, any.missing = FALSE)
Error: Assertion on 'ref_time_values' failed: Must have length >= 1, but has length 0.

r$> ref_time_values = c(as.Date("2020-03-06"),
     NA)

r$> assert_numeric(ref_time_values, 
    min.len = 1L, null.ok = FALSE, any.missing
     = FALSE)
Error: Assertion on 'ref_time_values' failed: Contains missing values (element 2).
```

dshemetov commented on R/grouped_epi_archive.R:
```diff
@@ -230,32 +209,23 @@ grouped_epi_archive <-
-        if (length(before) != 1L || is.na(before) || before < 0L) {
-          Abort("`before` must be length-1, non-NA, non-negative.")
-        }
+        assert_int(before, lower = 0L, null.ok = FALSE, na.ok = FALSE)
```
```R
r$> before = c(1, 2)

r$> assert_int(before, lower = 0L, null.ok = F
    ALSE, na.ok = FALSE)
Error: Assertion on 'before' failed: Must have length 1.

r$> assert_int(before, lower = 0L, null.ok = F
    ALSE, na.ok = FALSE)
Error: Assertion on 'before' failed: May not be NA.

r$> assert_int(before, lower = 0L, null.ok = F
    ALSE, na.ok = FALSE)
Error: Assertion on 'before' failed: Element 1 is not >= 0.
```

dshemetov commented on R/grouped_epi_archive.R:
```diff
@@ -265,15 +235,9 @@ grouped_epi_archive <-
-        }
+        assert_logical(as_list_col, len = 1L)
+        assert_logical(all_versions, len = 1L)
+        assert_character(names_sep, len = 1L, null.ok = TRUE)
```
See examples for these above (messaging is similar).

dshemetov commented on R/grouped_epi_archive.R:
```diff
@@ -290,9 +254,7 @@ grouped_epi_archive <-
-          if (!(is.atomic(comp_value) || is.data.frame(comp_value))) {
-            Abort("The slide computation must return an atomic vector or a data frame.")
-          }
+          assert(check_atomic(comp_value, any.missing = TRUE), check_data_frame(comp_value), combine = "or", .var.name = vname(comp_value))
```
```R
r$>           assert(check_atomic(comp_value, 
    any.missing = TRUE), check_data_frame(comp
    _value), combine = "or", .var.name = vname
    (comp_value))
Error: Assertion on 'comp_value' failed: One of the following must apply:
 * check_atomic(comp_value): Must be of
 * type 'atomic', not 'list'
 * check_data_frame(comp_value): Must be
 * of type 'data.frame', not 'list'.
```

dshemetov commented on R/slide.R:
```diff
@@ -168,55 +168,43 @@ epi_slide <- function(x, f, ..., before, after, ref_time_values,
+    )
+    if (anyDuplicated(ref_time_values) != 0L) {
+      cli_abort("`ref_time_values` must not contain any duplicates; use `unique` if appropriate.")
+    }
```
See similar section in `grouped_epi_archive` above.

dshemetov commented on R/slide.R:
```diff
@@ -168,55 +168,43 @@ epi_slide <- function(x, f, ..., before, after, ref_time_values,
-    if (length(before) != 1L || is.na(before) || before < 0L) {
-      Abort("`before` must be length-1, non-NA, non-negative")
-    }
+    assert_int(before, lower = 0, null.ok = FALSE, na.ok = FALSE)
```
Default messages are similar, see above.

dshemetov commented:
@dajmcdon I've updated this PR with better messaging in certain cases. I also went through all the instances where I changed to using a checkmate assert and added an example of its error message. I think the remaining ones are all very similar to what we had before. Let me know what you think!

dajmcdon reviewed (approved):
I added one comment here: https://github.com/cmu-delphi/epiprocess/pull/413#discussion_r1480763988.

If you're fine with it then so am I.

dshemetov commented:
Thanks! I'm happy to polish a little more. Just let me know what you'd prefer to see that in that case.

dshemetov commented:
Going to merge for now, but happy to polish messages in another PR.
